core/nodejsActionBase/src/service.js

```diff
@@ -44,112 +65,96 @@ function NodeActionService(config, logger) {
+     * (the user code threw an exception, or our proxy had an internal error).
+     * Actions returning { error: ... } are modeled as a Promise successful resolution.
+     *
      * req.body = { value: Object, meta { activationId : int } }
      */
-    this.runCode = function runCode(req, res) {
+    this.runCode = function runCode(req) {
         if (status === Status.ready) {
             setStatus(Status.running);
-            doRun(req, function(result, errMsg) {
+
+            return doRun(req).then(function (result) {
                 setStatus(Status.ready);
-                if (result) {
-                    // it is an error if the value is defined but not of type object
-                    if (typeof result !== 'object') {
-                        console.error('Result must be of type object but has type "' + typeof result + '":', result);
-                        res.status(502).json({ error: 'The action did not return a dictionary.' });
-                    } else {
-                        res.status(200).json(result);
-                    }
+
+                if (typeof result !== "object") {
+                    console.error('Result must be of type object but has type "' + typeof result + '":', result);
+                    return errorMessage(502, "The action did not return a dictionary.");
                 } else {
-                    res.status(500).json({ error: 'An error has occurred: ' + errMsg });
+                    return responseMessage(200, result);
                 }
+            }).catch(function (error) {
+                setStatus(Status.ready);
+
+                return Promise.reject(errorMessage(500, "An error has occurred: " + error));
             });
-        } else if (status === Status.starting || status === Status.running) {
-            logger.info('[runCode]', 're-scheduling runCode due to status',  status);
-            setTimeout(function() { runCode(req, res); }, 1); // Node can achieve 1 ms
         } else {
             logger.info('[runCode]', 'cannot schedule runCode due to status', status);
-            res.status(500).json({});
-        }
-    }
-
-    function shutdown(res, code, msg) {
-        status = Status.stopped
-        logger.info('[setStatus]', 'stopping now');
-        res.status(code).json(msg);
-        server.close();
-    }
-
-    function setStatus(newStatus) {
-        if (status !== Status.stopped) {
-            status = newStatus;
+            return Promise.reject(errorMessage(500, "Internal system error: container not ready."));
         }
     }
 
     function doInit(message) {
         var context = newWhiskContext(config, logger);
-        userScript = new NodeActionRunner(message, context);
-        if (typeof userScript.userScriptMain === 'function') {
+
+        userCodeRunner = new NodeActionRunner(context);
+
+        return userCodeRunner.init(message).then(function (result) {
```

Comment:
Same comment here about `result`. It isn't used here and the value coming from `NodeActionRunner.init()` isn't really meaningful anyway.